### ISSUE TITLE
presentPaymentSheet X to close not returning as cancelled

### ISSUE THREAD

**Issue by drewbietron** [stripe/stripe-react-native#1302]
**Describe the bug**
When a saved card is selected in the `presentPaymentSheet` UI, you can click the top left 'X' to dismiss the payment sheet, or click outside of the payment sheet but no error will be returned.  This means if you are relying on the error to know whether to call `confirmPaymentSheetPayment` - you will be attempted a charge on the saved card even though the user dismissed the payment sheet in the UI.  The also happens is Link is selected the same as your saved card.

```typescript
const { error } = await presentPaymentSheet();
// error will not return anything in the example

// If no error is found, 
const { error: confirmPaymentSheetError } = await confirmPaymentSheetPayment();

// This runs since there's no error returned with presentPaymentSheet, 😳 
// Obviously that's no bueno, users are confirming payments when they are clearly trying to cancel.
```

**To Reproduce**
1) Have a saved card 
2) Click the top left X in the payment sheet UI or click outside of the payment sheet modal to dismiss the payment sheet
3) `presentPaymentSheet` does not return an error and there's no way to know 

**Expected behavior**
`presentPaymentSheet` should return a cancelled error

**Screenshots**
Attached a video.  First video shows how its not returning an error for cancelled with the saved card selected.  After removing the card, `presentPaymentSheet` returns an error as cancelled.

While clicking the X
https://user-images.githubusercontent.com/1255104/219280903-c174a944-06fa-4631-a64a-ba514ffa5783.mp4

While clicking outside of the payment sheet modal
https://user-images.githubusercontent.com/1255104/219282028-8eae324b-76d8-4a57-922f-432fd05b0b07.mp4

With no saved card or payment method in the modal selected, the error is returned
https://user-images.githubusercontent.com/1255104/219281091-1c7b11a0-fbf3-4e1a-ba21-1e230fbf99ed.mp4

**Comment by charliecruzan-stripe** [comment]
heya @drewbietron - it looks like you're using the custom flow for Payment Sheet, but the regular flow would serve you better and avoid this problem.

Right now, you're passing `customFlow: true` to `initPaymentSheet`. This separates out the payment method selection & payment confirmation steps. But you're calling `confirmPaymentSheetPayment` in code after `presentPaymentSheet`, so really you can just get rid of `customFlow: true`, and `confirmPaymentSheetPayment`.


The custom flow is meant for a case when, for one example, you have two buttons:
- one for selecting the payment method
- one for "checkout" where you actually call `confirmPaymentSheetPayment`

Let me know if that clears things up or if i'm missing something

**Comment by drewbietron** [comment]
> heya @drewbietron - it looks like you're using the custom flow for Payment Sheet, but the regular flow would serve you better and avoid this problem.
> 
> Right now, you're passing `customFlow: true` to `initPaymentSheet`. This separates out the payment method selection & payment confirmation steps. But you're calling `confirmPaymentSheetPayment` in code after `presentPaymentSheet`, so really you can just get rid of `customFlow: true`, and `confirmPaymentSheetPayment`.
> 
> The custom flow is meant for a case when, for one example, you have two buttons:
> 
> * one for selecting the payment method (this is where you call `presentPaymentSheet`)
> * one for "checkout"  (this is where you call `confirmPaymentSheetPayment`)
> 
> Let me know if that clears things up or if i'm missing something

Thanks for the reply @charliecruzan-stripe 

I am using the custom flow so that I can verify the checkout is valid before charging for the payment.  The particular use case is for purchasing items that may be out of stock at time of purchase and between the presentPaymentSheet and actually capturing the charge, it is likely that duration of time would cause an item to potentially be out of stock.  I want to ensure the items are in stock before calling `confirmPaymentSheetPayment`.  Do I have any options here?

**Comment by charliecruzan-stripe** [comment]
How long is the duration of time that an item could go out of stock? We'll be adding a timeout feature to `presentPaymentSheet` in https://github.com/stripe/stripe-react-native/pull/1287, do you think that would help?

**Comment by drewbietron** [comment]
> How long is the duration of time that an item could go out of stock in? We'll be adding a timeout feature to `presentPaymentSheet` in https://github.com/stripe/stripe-react-native/pull/1287, do you think that would help?
> 
> 
> 
> Let me see if this is expected behavior for `presentPaymentSheet`, your use case does make sense to me

This use case is for items that are announced to be "dropped" at a certain time and there's lots of buyers trying to purchase them at the same time. Depending on how long it takes for each buyer to go through the payment sheet details, someone is going to end up not getting the items. If the custom flow worked with cancelling the presented payment sheet I think we'd be all set. For now if I get rid of the custom flow I'll be refunding all of the orders whose payments were captured for out of stock items.

**Comment by charliecruzan-stripe** [comment]
> cancelling the presented payment sheet doesn't throw a cancelled error

It returns a `paymentOption` if one is selected in the UI at the time of cancellation, similar to how `initPaymentSheet` returns a `paymentOption` if one exists for the customer. 

Then you take that response and can use it to indicate to your customer that this is their "pre selected" payment option. If there is no payment option selected, then the flow has been "cancelled", but if there is a payment option selected, then that one is returned

**Comment by drewbietron** [comment]
> > cancelling the presented payment sheet doesn't throw a cancelled error
> 
> 
> 
> It returns a `paymentOption` if one is selected in the UI at the time of cancellation, similar to how `initPaymentSheet` returns a `paymentOption` if one exists for the customer. 
> 
> 
> 
> Then you take that response and can use it to indicate to your customer that this is their "pre selected" payment option. If there is no payment option selected, then the flow has been "cancelled", but if there is a payment option selected, then that one is returned
> 
> 
> 
> > Is there something else I can look at from the response of the presented payment sheet to know if it's safe to call confirmPaymentSheet method?
> 
> 
> 
> The `paymentOption`

Thanks. 

Is there a way in this case to know if it's a preselected option and the modal was dismissed though?  Is it sane to expect the modal to return as cancelled if the user doesn't actually choose to proceed with the select payment?  If there's a way to know that the user didn't proceed to confirm the selected payment method via payment option that seems viable, but it sounds like regardless of choosing to continue with the selected payment, or cancelling the modal, the response is the same?

Appreciate your responses on this.

**Comment by charliecruzan-stripe** [comment]
No, there isn't currently a way to know if a payment option is selected but the sheet was dismissed.

> it sounds like regardless of choosing to continue with the selected payment, or cancelling the modal, the response is the same

Correct, because the user has a selected `paymentOption` in either case. Out of curiosity- how would you handle that case differently in your checkout flow? Simply have the user have to revisit the "select payment method" flow and **not** cancel?

**Comment by drewbietron** [comment]
> No, there isn't currently a way to know if a payment option is selected but the sheet was dismissed.
> 
> > it sounds like regardless of choosing to continue with the selected payment, or cancelling the modal, the response is the same
> 
> Correct, because the user has a selected `paymentOption` in either case. Out of curiosity- how would you handle that case differently in your checkout flow? Simply have the user have to revisit the "select payment method" flow and **not** manually dismiss it?
 
In the case which I was expecting - if they dismiss the modal, then an error is returned with a cancelled state, which would not allow for confirming the payment method via a custom flow.  They never intended to purchase, since they didn't actually choose to continue with whatever payment method is selected, so bail.

If the user didn't proceed via selecting the option to proceed with a saved payment method (eg the modal is closed) they would just reopen the modal via presentPaymentSheet and start the flow over.  They did close it for a reason 🤷.  The flow from presentPaymentSheet is only continued if they actually proceed with a payment method.  The same as if they didn't have a saved payment method and closed the modal which works as expected atm (albeit because theres no payment method to proceed with)

This only happens when a customer has a saved payment method (or Link in the case where its avaiable, which is available to users without a saved payment method) and I'd also like to point out that the selected payment method is selected by default if its saved for the customer.  So there's no way to unselect it unless the user deletes the saved payment method, or chooses to add a new one and dismisses the modal during the flow.

In summary, its the point that a user with a saved payment method can present the modal sheet, and then close the modal without actually choosing to proceed with that chosen payment method and still be charged since there is no programatic way atm to know if the user has explicitly chosen to proceed, or to close the modal.

**Comment by charliecruzan-stripe** [comment]
Choosing the selected payment method and then proceeding with confirming the payment are two separate steps in the custom flow for Payment Sheet, so a user _can_ close the modal without actually proceeding with being charged, but you have to implement those two steps separately. [In the case i used as an example above](https://github.com/stripe/stripe-react-native/issues/1302#issuecomment-1433645935), this means two user actions: the first being selecting their payment method, and the second being "ok, now i'd like to pay using that payment method". This way, if a user opens the payment method selection sheet, and cancels (even though they still have a paymentOption), they don't **need** to click the "buy" button.

**Comment by charliecruzan-stripe** [comment]
Closing this since this is expected behavior on our end- when using the custom flow, the user's currently selected payment option will be returned by `presentPaymentSheet` when the sheet is manually closed (this is consistent on both platforms)

**Comment by abdullahIsa** [comment]
Hmm having same kind of issue too and sometimes the flow of payment is such that `confirmPaymentSheetPayment()` is called after user has inputted their card details and clicked to pay but if they come back and their card details already saved and they decided "oh let me cancel" but they got charged, one thing of course is `presentPaymentSheet()` will return {error, paymentOption} with `paymentOption` containing selected card info and i can show a modal to user if they want to pay since a card is selected but this is kinda bad flow i would expect `presentPaymentSheet()` to return "canceled" when user clicked cancel icon or `presentPaymentSheet()` background or some sort of callBack when cancel icon is clicked or background is clicked.

It would really be appreciative if that can be added than having to make two different buttons or some sort of modal just to inform user that intentionally wants to cancel that they want to cancel or proceed with payment, thanks.

**Comment by charliecruzan-stripe** [comment]
@abdullahIsa what's the use case for using the custom flow for payment sheet if you are calling `confirmPaymentSheetPayment` directly after (without a confirmation step required of the user)?

**Comment by abdullahIsa** [comment]
> @abdullahIsa what's the use case for using the custom flow for payment sheet if you are calling `confirmPaymentSheetPayment` directly after (without a confirmation step required of the user)?

I am the creator of https://chappie.app a multi tool app, one of the tools is wallet and in one of its section is the top up, user decides amount, fills up any needed details and clicks a single button "top up" and a strip paymenSheet will show up, now lets say user has done top up before so their card would be saved by strip and now lets say paymentSheet lists all the saved card and user clicked one, user would expect all the process to be single as they already clicked the card to top up with and with that i call `confirmPaymentSheetPayment` along the process to charge the user but now let say user sees all listed cards and decides to click cancel icon but strip decides to consider it as user clicked one of the saved cards which would trigger the payment flow of charging user when doing "top up", of course like i said i could show an alert such as "are you sure you want to top up" but i find that naive as user intention is to cancel not continue with the process.

I hope understanding.
